### PR TITLE
refactor!: Remove/replace ParameterSet-based measurement (EDM rework 4d/4)

### LINKED ISSUE #350: Acts::Measurement infrastructure has unacceptable build overhead
After trimming a bunch of Eigen fat, the compile-time overhead of Acts' fitting unit tests and examples becomes bottlenecked by the Measurement infrastructure, and more precisely the "unique ordered sublists" variant that it provides. To quantify this, about 50% of template-related compilation time for these compilation units is spent processing that code alone.

---

Given an N-dimensional state space `(x1, x2, ..., xN)`, this mechanism provides an `std::variant` that goes over all possible tuples `std::variant<(x1), (x1, x2), (x1, x3), ..., (x1, xN), (x2), (x2, x3), ..., (x2, xN), (x1, x2), (x1, x2, x3), ..., (x1, x2, xN), ..., (x1, x2, ..., xN)>`.

As you can imagine, the combinatorics are not in our favor here, and the resulting `std::variant` has a lot of different alternatives. More precisely, since this variant is semantically equivalent to an `(std::optional<x1>, std::optional<x2>, ..., std::optional<xN>)` tuple with the additional constraint of excluding the "all optionals empty" null state, we can easily compute that the number of variants will be 2^N-1, which for the N=6 case that Acts is currently using means 63 different variants.

The overhead associated with generating this type and its basic methods is already quite significant in practice because libstdc++'s implementation of `std::variant` is not optimized for low compile-time overhead, but for high-quality error reporting. It sends the input types through various type traits for SFINAE and static assertions, and type traits themselves are not super cheap to instantiate.

Furthermore, the overhead is not restricted to the instantiation of that `std::variant` and its basic operations like the constructor. Every `std::visit` through that `std::variant` will have its associated code duplicated 63 times. So for example, templated code which has a RAM footprint of 10MB when instantiated once will get, in the worst case where the compiler does not manage to deduplicate anything, a RAM footprint of 630MB when fed through `std::visit`.

---

During my last conversation with @paulgessinger about this mechanism, I did not manage to understand why exactly this std::variant cannot be replaced with `std::tuple<std::optional<x1>, std::optional<x2>, ..., std::optional<xN>>`, or an optimized version thereof.

In general, I do not currently understand this mechanism well enough to see if we actually need it, let alone optimize it. All I can say, from my current perspective, is that getting rid of it, if at all possible, would likely be highly beneficial to Acts' compilation times and build RAM consumption.

@paulgessinger and @msmk0 have expressed interest in working on this. I wouldn't mind helping either, but I need to understand this better to work on it, so if I were to work on this someone first needs to sit down with me for a while until I fully get it.

See also #280 where Paul proposes an alternate design for both Measurement and MultiTrajectory.

### PR DESCRIPTION
**Pull request by msmk0**
BREAKING CHANGE: `Measurement` is replaced with a new implementation that is templated only on the measurement dimension and does not store a reference object anymore.`Fittable{Volume}Measurement` is replaced by new variants `{Bound,Free}VariantMeasurement` based on the updated `Measurement` type. `ParameterSet` and related types are removed. Since the `Measurement` does not store the reference object anymore, setting the calibrated measurement for a `MultiTrajectory` state does not automatically set the reference surface anymore. The reference surface must always be set explicitly.

In addition to the core changes, this includes quite a few dependent changes in `Fatras`, `Examples`, and `Tests` to adapt to the new measurement types:

- Rewrite the `Fatras` smearing digitizers. The are now also templated only the the measurement size with the measurement subspace configurable at runtime.
- Rewrite the smearing configuration in the `Examples` to match the updated `Fatras` types.
- `Examples` algorithms that handle measurements and require access to the local-to-global transforms now take require a `TrackingGeometry` pointer via their configuration structs. It is required since the `Surface` is not available anymore directly via the measurements.
- The `TestSourceLink` in the unit tests now stores the measurement data inline. Previously, the `TestSourceLink` just stored a pointer to a measurement which results in a circular dependency, i.e. `TestSourceLink` points to `FittableMeasurement` which stores a copy of `TestSourceLink`. This worked before by storing fake information in at least one of the objects involved. It does not work anymore with the updated `Measurement` type. The updated `TestSourceLink` also matches the intended data flow for source links, i.e. you have the source link first and then use a calibrator to retrieve the `Measurement`.

Contains #554 and #555 and must be merged afterwards. Finally closes #350.

### CODE REVIEW

**Comment by msmk0** [comment]
Not done/ready. I opened this PR mostly to track my progress for now.

File: Core/include/Acts/EventData/Measurement.hpp [hunk 1]
```diff
@@ -9,387 +9,244 @@
...
 ///
-/// If the measurement is in local parameters and will not provide localToGlobal
-/// information. It is thus free from any Context.
+/// The measurement is templated on the measurement subspace size and its
+/// content (indices) is defined at runtime. In previous implementations,
+/// the measurement was templated directly on the subspace indices. This lead to
+/// a combinatorial explosion and severe penalty during compilation since all
+/// possible combinations need to be considered for the measurement variant.
+/// However, for the computations only the size of the measurement subspace is
+/// relevant.
```
Comment (asalzburger) [inline Core/include/Acts/EventData/Measurement.hpp:36]: I would remove the mention of previous implementations, I suppose we should keep up with documenting the code as is and not as it was?

File: Core/include/Acts/EventData/Measurement.hpp [hunk 2]
```diff
@@ -9,387 +9,244 @@
...
+/// The measurement intenionally does not store a pointer/reference to the
+/// explicit reference object in the geometry hierarchy, i.e. the surface or
+/// volume. The reference object can already be identified via the geometry
+/// identifier provided by the source link. Since a measurement
+/// **must** be anchored within the geometry hierarchy, all measurement surfaces
+/// and volumes **must** provide valid geometry identifiers. In all use-cases,
+/// e.g. Kalman filtering, a pointer/reference to the reference object is
+/// available before the measurement is accessed; e.g. the propagator provides
+/// the surface pointer during navigation, which is then used to lookup possible
+/// measurements.
```
Comment (asalzburger) [inline Core/include/Acts/EventData/Measurement.hpp:47]: Here's one concern I am having: what if someone wants to express a bunch of measurements (e.g. a muon track) on a chosen surface that is NOT part of the hierarchy? 
The Kalmanfilter will allow this with the `external measurements` by what do we do here? do we need an extra container?

File: Core/include/Acts/EventData/detail/CalculateResiduals.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    // regular, unrestricted parameter.
+    if (fullIndex == eBoundPhi) {
+      residuals[i] = difference_periodic<OutputScalar>(
+          measured[i], reference[fullIndex],
+          static_cast<OutputScalar>(2 * M_PI));
+    } else {
+      residuals[i] = measured[i] - reference[fullIndex];
+    }
+  }
+}
```
Comment (asalzburger) [inline Core/include/Acts/EventData/detail/CalculateResiduals.hpp:64]: Ok, that gets so much easier to read. 
The reason why you can't call `Eigen::colwise()` here is that `phi` needs special treatment, right?

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/Smearers.hpp [hunk 4]
```diff
@@ -12,10 +12,11 @@
...
 #include "ActsFatras/Digitization/DigitizationError.hpp"
-#include <climits>
+
 #include <cmath>
-#include <exception>
+#include <limits>
 #include <random>
+#include <utility>
 
 namespace ActsExamples {
```
Comment (asalzburger) [inline Examples/Algorithms/Digitization/include/ActsExamples/Digitization/Smearers.hpp:21]: Has this file gone entirely ?

Comment (asalzburger) [reply Examples/Algorithms/Digitization/include/ActsExamples/Digitization/Smearers.hpp:21]: Ah, I see, just the unrolling of the SFGenerator is deleted, not needed anymore.

File: Examples/Algorithms/Digitization/src/DigitizationOptions.cpp [hunk 5]
```diff
@@ -47,11 +48,48 @@ void ActsExamples::Options::addDigitizationOptions(
   opt("digi-smear-parameters",
       value<read_range>()->multitoken()->default_value({}),
       "Smearing Input: smear parameters depending on the smearing type, 1 "
-      "parameter for simple gauss, 4 for all others (1 parameter, 2 absolute "
+      "parameter for simple gauss, 3 for all others (1 parameter, 2 absolute "
       "range "
       "values, 1 range indicator: both neg < 0, pos/neg = 0, both pos > 0).");
```
Comment (asalzburger) [inline Examples/Algorithms/Digitization/src/DigitizationOptions.cpp:53]: Don't we need the fourth one anymore? I can not have a negative value in the read_series, which is why I needed the range indicator

Comment (asalzburger) [reply Examples/Algorithms/Digitization/src/DigitizationOptions.cpp:53]: Anyhow, we really need that `.json` configured.

File: Examples/Io/Root/include/ActsExamples/Io/Root/RootDigitizationWriter.hpp [hunk 6]
```diff
@@ -124,28 +125,8 @@ class RootDigitizationWriter : public WriterT<MeasurementContainer> {
...
-      if (ps.template contains<Acts::eBoundTheta>()) {
-        tree->Branch(std::string("rec_" + bNames[Acts::eBoundTheta]).c_str(),
-                     &recBound[Acts::eBoundTheta]);
-      }
-      if (ps.template contains<Acts::eBoundTime>()) {
-        tree->Branch(std::string("rec_" + bNames[Acts::eBoundTime]).c_str(),
-                     &recBound[Acts::eBoundTime]);
-      }
+    void setupBoundRecBranch(Acts::BoundIndices i) {
+      tree->Branch(std::string("rec_" + bNames[i]).c_str(), &recBound[i]);
```
Comment (asalzburger) [inline Examples/Io/Root/include/ActsExamples/Io/Root/RootDigitizationWriter.hpp:129]: Ok, this clearly wins.

**Review by asalzburger — APPROVED** [review]
Amazing work, @msmk0.

The one thing we will have to deal with in the future, if the story of unassociated measurements (see my comment), but we can always have them in a separate container that is not the tracking geometry.

File: Core/include/Acts/EventData/detail/CalculateResiduals.hpp (hunk 3 above)
Comment (msmk0) [reply Core/include/Acts/EventData/detail/CalculateResiduals.hpp:64]: Yes, and the fact that I need to do the subspace selection, i.e. I do not need to access every element.

File: Examples/Algorithms/Digitization/src/DigitizationOptions.cpp (hunk 5 above)
Comment (msmk0) [reply Examples/Algorithms/Digitization/src/DigitizationOptions.cpp:53]: I think, this was a documentation error to begin with. All smearers that are available take three parameters, not four (simple Gaussian excluded).

File: Core/include/Acts/EventData/Measurement.hpp (hunk 1 above)
Comment (msmk0) [reply Core/include/Acts/EventData/Measurement.hpp:37]: Ok, I will update the documentation in #555.

File: Core/include/Acts/EventData/Measurement.hpp (hunk 2 above)
Comment (msmk0) [reply Core/include/Acts/EventData/Measurement.hpp:48]: I think that needs a bit more thought anyways. So far, all measurements **must** be anchored in the geometry hierarchy. Both the KF and the CKF use the navigator to provide e.g. the surface and then use the geometry identifier to lookup possible measurements (in the CKF case). This was already a requirement before this PR and removing reference object from the measurement only removes duplicate information (and thus greatly simplifies handling both surface- and volume-based without relying on complicated template hacks).

If you want to support external measurements in the KF that are not anchored in the geometry hierarchy, you would probably need to add support to the navigator anyways. In which case you could provide the external measurements as pairs of reference object and measurement.

**Comment by msmk0** [comment]
@asalzburger @paulgessinger Rebased and ready for a review.

File: Examples/Algorithms/Digitization/src/DigitizationOptions.cpp (hunk 5 above)
Comment (asalzburger) [reply Examples/Algorithms/Digitization/src/DigitizationOptions.cpp:53]: No, indeed the parameters need 3, but for the command line I need 4.

Comment (msmk0) [reply Examples/Algorithms/Digitization/src/DigitizationOptions.cpp:53]: Ah, no I understand. Will fix that.

File: Core/include/Acts/EventData/Measurement.hpp (hunk 2 above)
Comment (asalzburger) [reply Core/include/Acts/EventData/Measurement.hpp]: > 
> If you want to support external measurements in the KF that are not anchored in the geometry hierarchy, you would probably need to add support to the navigator anyways. In which case you could provide the external measurements as pairs of reference object and measurement.

This is what @Corentin-Allaire  is working on, actually.

File: Core/include/Acts/EventData/detail/CalculateResiduals.hpp (hunk 3 above)
Comment (asalzburger) [reply Core/include/Acts/EventData/detail/CalculateResiduals.hpp]: Understood.

File: Core/include/Acts/EventData/Measurement.hpp (hunk 1 above)
Comment (asalzburger) [reply Core/include/Acts/EventData/Measurement.hpp]: Merci.

File: Examples/Algorithms/Digitization/src/DigitizationOptions.cpp (hunk 5 above)
Comment (asalzburger) [reply Examples/Algorithms/Digitization/src/DigitizationOptions.cpp:53]: Unfortunately, this shows the limits of the command line implementation.

Comment (msmk0) [reply Examples/Algorithms/Digitization/src/DigitizationOptions.cpp]: @asalzburger Should be fixed now.

**Review by asalzburger — APPROVED** [review]
The changes from my last review to now look good, I am just compiling the branch and checking if the digitisation works as an example.

**Comment by asalzburger** [comment]
Second finding:
- even then, the smearers do not seem to be correctly configured. I am chasing the following error:

```shell
   49  	    throw std::invalid_argument("Missing random numbers tool");
   50  	  }
   51  	  if (m_cfg.smearers.empty()) {
-> 52  	    throw std::invalid_argument("Missing smearers configuration");
   53  	  }
   54  	  // create the smearers from the configuration
   55  	  std::vector<std::pair<Acts::GeometryIdentifier, Smearer>> smearersInput;
(lldb) print m_cfg
(ActsExamples::SmearingAlgorithm::Config) $0 = {
  inputSimHits = "hits"
  outputSourceLinks = "sourcelinks"
  outputMeasurements = "measurements"
  outputMeasurementParticlesMap = "measurement_particles_map"
  outputMeasurementSimHitsMap = "measurement_simhits_map"
  trackingGeometry = std::__1::shared_ptr<const Acts::TrackingGeometry>::element_type @ 0x00000001098a65e0 strong=6 weak=1 {
    __ptr_ = 0x00000001098a65e0
  }
  randomNumbers = std::__1::shared_ptr<const ActsExamples::RandomNumbers>::element_type @ 0x0000000102527628 strong=6 weak=1 {
    __ptr_ = 0x0000000102527628
  }
  smearers = {
    m_ids = size=0 {}
    m_masks = size=0 {}
    m_values = size=0 {}
  }
}
```

**Comment by msmk0** [comment]
@asalzburger The PR now includes your fixes from #616 .

**Review by asalzburger — APPROVED** [review]

**Comment by asalzburger** [comment]
Well down @msmk0 !!!

**Pull request merged**